### PR TITLE
Refactor zone like objects

### LINKED ISSUE #982: Beam is not rendering
image
There should be an arrow indicating where beam vector is located and what direction it is pointing but for some reason it disappeared.

### PR DESCRIPTION
**Pull request by Derstilon**
Closes #982 #983 
Merge util catalogs into one
Move all simulation related ThreeJS component to segratated cataloges in ThreeEditor/Simulation/
Simulation objects categorized to:

- Base: all abstract extension of Base ThreeJS objects
- Detectors: detector geometires sliced into bins
- Figures: all non material meshes defining areas in simulations
- Materials: Simulation materials and related
- Physics: Physics configuration, beam and related
- Scoring: output, filters and quantities
- SpecialComponents: Voxelized CTCube and any future compononts like beam modulators
- Zones: WorldZone, BooleanZone and any future components with geometry definition and physical material

CTMesh renamed to CTCube
BasicMesh and most inheriting classes renamed to ___Figure

image

### CODE REVIEW

**Comment by grzanka** [comment]
@Derstilon check if loading URL works: 
image
I made a test with [https://raw.githubusercontent.com/grzanka/plotly_lgad_dosimetry/main/montecarlo/shield-hit12a/results_bp.json](https://yaptide.github.io/web_dev/?https://raw.githubusercontent.com/grzanka/plotly_lgad_dosimetry/main/montecarlo/shield-hit12a/results_bp.json)

**Comment by grzanka** [comment]
Merging this PR would break rendering of my precalculated results, saved here:
https://github.com/grzanka/plotly_lgad_dosimetry/tree/main/montecarlo/shield-hit12a

Could you please make a fork of https://github.com/grzanka/plotly_lgad_dosimetry, rename what is needed in that JSON and make a PR with fix ?

**Review by grzanka — DISMISSED** [review]

**Comment by grzanka** [comment]
@Derstilon remember to bump JSON standard version

File: src/App.tsx [hunk 1]
```diff
@@ -10,6 +10,13 @@ import { ShSimulation } from './services/ShSimulatorService';
 import { Store } from './services/StoreService';
 import WrapperApp from './WrapperApp/WrapperApp';
 
+declare module '@mui/material/styles' {
+	// add new variables to the theme type
+	interface TypeBackground {
+		secondary?: string;
```
Comment (ostatni5) [inline src/App.tsx:16]: why this is optional?

File: src/ThreeEditor/Simulation/Base/SimElement.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import * as THREE from 'three';
+import { Editor } from '../../js/Editor';
+import { SimulationPropertiesType } from '../../../types/SimProperties';
+import { ISimulationSceneChild, SimulationSceneContainer } from './SimScene';
+
+export abstract class SimulationElement
```
Comment (ostatni5) [inline src/ThreeEditor/Simulation/Base/SimElement.ts:6]: Chnage file name to match it

File: src/ThreeEditor/Simulation/Base/SimElement.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import * as THREE from 'three';
+import { Editor } from '../../js/Editor';
+import { SimulationPropertiesType } from '../../../types/SimProperties';
+import { ISimulationSceneChild, SimulationSceneContainer } from './SimScene';
+
```
Comment (ostatni5) [inline src/ThreeEditor/Simulation/Base/SimElement.ts:5]: add comment

File: src/ThreeEditor/Simulation/Figures/BasicFigures.ts [hunk 4]
```diff
@@ -68,46 +68,21 @@ export class CylinderMesh extends BasicMesh<THREE.CylinderGeometry> {
...
+		super(editor, 'Sphere', 'SphereFigure', 'Sphere', geometry ?? sphereGeometry, material);
 	}
 }
 
-const ctGeometry = new THREE.BoxGeometry(2, 0.5, 1, 1, 1, 1);
-const ctMaterial = defaultMaterial.clone();
-ctMaterial.color.setHex(0x00ff00);
-export class CTMesh extends BasicMesh<THREE.BoxGeometry> {
-	readonly notScalable = true;
+export const isBasicMesh = (x: unknown): x is BasicFigure => x instanceof BasicFigure;
```
Comment (ostatni5) [inline src/ThreeEditor/Simulation/Figures/BasicFigures.ts:82]:
```suggestion
export const isBasicFigure = (x: unknown): x is BasicFigure => x instanceof BasicFigure;
```

File: src/ThreeEditor/Simulation/Figures/FigureScene.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import * as THREE from 'three';
+import { Editor } from '../../js/Editor';
+import { generateSimulationInfo } from '../../../util/AdditionalGeometryData';
+import { BasicFigure } from './BasicFigures';
+import { SimulationPropertiesType } from '../../../types/SimProperties';
+import { UniqueChildrenNames, getNextFreeName } from '../../../util/Name/Name';
+
+export class FigureScene
+	extends THREE.Scene
+	implements SimulationPropertiesType, UniqueChildrenNames
```
Comment (ostatni5) [inline src/ThreeEditor/Simulation/Figures/FigureScene.ts:10]: Fix to use Simulation scene

File: src/ThreeEditor/Simulation/Zones/WorldZone/WorldZoneHelper.ts [hunk 6]
```diff
@@ -23,14 +23,14 @@ export class WorldZoneHelper extends THREE.Object3D {
 			case 'Cylinder':
 				return this.cylinderMesh;
 			case 'Sphere':
-				return this.sphereMesh;
+				return this.SphereFigure;
```
Comment (ostatni5) [inline src/ThreeEditor/Simulation/Zones/WorldZone/WorldZoneHelper.ts:26]: Why there is Figure and mesh?

File: src/ThreeEditor/Simulation/Zones/ZoneManager.ts [hunk 7]
```diff
@@ -105,7 +107,7 @@ export class ZoneManager extends THREE.Scene implements ISimulationObject {
 	}
 
 	toJSON() {
-		const zones = this.zoneContainer.toJSON() as ZoneJSON[];
+		const zones = this.zoneContainer.toJSON() as BooleanZoneJSON[];
```
Comment (ostatni5) [inline src/ThreeEditor/Simulation/Zones/ZoneManager.ts:110]:
```suggestion
		const zones : BooleanZoneJSON[] = this.zoneContainer.toJSON();
```

File: src/ThreeEditor/components/Sidebar/properties/category/RangeModulatorConfiguration.tsx
Comment (ostatni5) [inline src/ThreeEditor/components/Sidebar/properties/category/RangeModulatorConfiguration.tsx:1]: obraz

File: src/ThreeEditor/js/commands/RemoveQuantityCommand.js [hunk 8]
```diff
@@ -1,49 +1,49 @@
-import { ScoringOutput } from '../../util/Scoring/ScoringOutput';
-import { ScoringQuantity } from '../../util/Scoring/ScoringQuantity';
+import { ScoringOutput } from '../../Simulation/Scoring/ScoringOutput';
+import { ScoringQuantity } from '../../Simulation/Scoring/ScoringQuantity';
 import { Command } from '../Command.js';
 // eslint-disable-next-line no-unused-vars
 import { Editor } from '../Editor.js';
```
Comment (ostatni5) [inline src/ThreeEditor/js/commands/RemoveQuantityCommand.js:5]:
```suggestion
// eslint-disable-next-line no-unused-vars
import type { Editor } from '../Editor.js';
```

File: src/ThreeEditor/js/commands/SetDetectTypeCommand.js [hunk 9]
```diff
@@ -22,9 +22,10 @@ export class SetDetectTypeCommand extends Command {
 
 	execute() {
 		this.object.detectType = this.newType;
-
+		console.log('SetDetectTypeCommand', this.object, this.newType);
```
Comment (ostatni5) [inline src/ThreeEditor/js/commands/SetDetectTypeCommand.js:25]:
```suggestion
```

File: src/types/SimProperties.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Type for properties of a simulation object.
+ * isObject3D is always true and is defined in the type because you can't implement a type in a class without defining at least one property.
+ * all other properties are optional and are used to properly handle the object in the editor.
+ */
+export type SimulationPropertiesType = {
+	readonly isObject3D?: true;
```
Comment (ostatni5) [inline src/types/SimProperties.ts:7]: Update to simobject

**Review by ostatni5 — DISMISSED** [review]
Adress comments

File: src/ThreeEditor/Simulation/Figures/FigureScene.ts (hunk 5 above)
Comment (Derstilon) [reply src/ThreeEditor/Simulation/Figures/FigureScene.ts:10]: It requires changes in many files that are already due to rewrite so will fix it later

**Review by grzanka — APPROVED** [review]

**Pull request merged**